### DISCUSSION TITLE
[Moveit! 2] Function definitions in header files.

### DISCUSSION THREAD

**Discussion by ibrahiminfinite** [moveit/moveit2#2073] [General]
This spawned off from #2018 

While going through [robot_state.h](https://github.com/ros-planning/moveit2/blob/main/moveit_core/robot_state/include/moveit/robot_state/robot_state.h) .
I noticed that it contains a good number of definitions.
At first I though this was mostly for const return functions, but then realized that was not the case.
This pattern is also present in other header files.

AndyZe suggests the following rule of thumb.
> As a rule of thumb for robot_state.h, perhaps we can say:
  If a function definition is 2 lines or less, put it in robot_state.h
  If a function definition is >2 lines, put it in robot_state.cpp
  
However, after some more thought , using the number of lines as a metric does not seem appropriate since it does not serve as a good  basis for reasoning other than for executable size, and so I am starting a discussion to understand more about this pattern.

There are function definitions in many `header` files throughout MoveIt 2,  however I am not able to come to a conclusion on why those definitions have been put into the `header` files rather than the respective `source`  files.

1. What is the reason for this, is this a deliberate choice or just happened to be like that ?
2. What is a good criteria to decide if a definition should be included in the `header` ?
3. Should definitions be included in `headers` at all ? Would it be better to keep the `headers` as declaration only ?

My personal opinion is to keep `headers` as declaration only as my reason for looking at the `headers` would be for finding out the available interfaces.

What are your thoughts on this matter @AndyZe , @sea-bass  , @tylerjw , @henningkayser.

**Comment by ibrahiminfinite** [comment]
A comment  about this by @henningkayser (from https://github.com/ros-planning/moveit2/issues/2018#issuecomment-1505367571)
> About definitions: I think any function blocks should go into source files. Some exceptions may be (in my opinion) function overloads that call each other in a single line and trivial getters. But even those would not be wrong to have in a source file.

**Comment by tylerjw** [comment]
Agree this thing is a mess. I'm happy to accept PRs that clean some of this up. I also agree that source files should contain the implementations.
